balancer/rls/internal/cache/cache.go

```diff
@@ -0,0 +1,237 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package cache provides an LRU cache implementation to be used by the RLS LB
+// policy to cache RLS response data.
+package cache
+
+import (
+	"container/list"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/internal/backoff"
+	"google.golang.org/grpc/status"
+)
+
+// Key represents the cache key used to uniquely identify a cache entry.
+type Key struct {
+	// Path is the full path of the incoming RPC request.
+	Path string
+	// KeyMap is a stringified version of the RLS request keys built using the
+	// RLS keyBuilder. Since map is not a Type which is comparable in Go, it
+	// cannot be part of the key for another map (the LRU cache is implemented
+	// using a native map type).
+	KeyMap string
+}
+
+// Entry wraps all the data to be stored in a cache entry.
+type Entry struct {
+	// Mu synchronizes access to this particular cache entry. The LB policy
+	// will also hold another mutex to synchornize access to the cache as a
+	// whole. To avoid holding the top-level mutex for the whole duration for
+	// which one particular cache entry is acted up, we use this entry mutex.
+	Mu sync.Mutex
+	// ExpiryTime is the absolute time at which the data cached as part of this
+	// entry stops being valid. When an RLS request succeeds, this is set to
+	// the current time plus the max_age field from the LB policy config. An
+	// entry with this field in the past is not used to process picks.
+	ExpiryTime time.Time
+	// BackoffExpiryTime is the absolute time at which an entry which has gone
+	// through backoff stops being valid.  When an RLS request fails, this is
+	// set to the current time plus twice the backoff time. The cache expiry
+	// timer will only delete entries for which both ExpiryTime and
+	// BackoffExpiryTime are in the past.
+	BackoffExpiryTime time.Time
+	// StaleTime is the absolute time after which this entry will be
+	// proactively refreshed if we receive a request for it. When an RLS
+	// request succeeds, this is set to the current time plus the stale_age
+	// from the LB policy config.
+	StaleTime time.Time
+	// BackoffTime is the absolute time at which the backoff period for this
+	// entry ends. The backoff timer is setup with this value. No new RLS
+	// requests are sent out for this entry until the backoff period ends.
+	BackoffTime time.Time
```

Comment:
How will this field be used?  It seems potentially redundant with `BackoffState.Timer`.

Reply:
No, that sounds fine.